### PR TITLE
[Messenger] Routing & Inheritance

### PR DESCRIPTION
**Pull request by sylfabre**
This PR details how Messenger deals with messages and inheritance.

We were discussing with my colleagues and had to do a quick POC to find out as it is not documented, and we had some doubt about it.

### CODE REVIEW

File: messenger.rst [hunk 1]
```diff
@@ -431,6 +431,35 @@ transport and "sending" messages there to be handled immediately:
             ],
         ]);
 
+Message Inheritance & Routing Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With the following code and configuration, messages will be dispatched like this:
+* `Notification` to `async_1`
+* `SmsNotification` to `async_1`
+* `SmsNotification` to `async_2`
```
Comment (OskarStark) [inline messenger.rst]:
```suggestion
* ``Notification`` to ``async_1``
* ``SmsNotification`` to ``async_1``
* ``SmsNotification`` to ``async_2``
```

File: messenger.rst [hunk 2]
```diff
@@ -431,6 +431,35 @@ transport and "sending" messages there to be handled immediately:
...
+
+.. code-block::php
+
+    <?php
```
Comment (OskarStark) [inline messenger.rst]:
```suggestion
```

File: messenger.rst [hunk 3]
```diff
@@ -431,6 +431,35 @@ transport and "sending" messages there to be handled immediately:
...
+    # config/packages/messenger.yaml
+    framework:
+        messenger:
+            routing:
+                App\Message\Notification: async_1
+                App\Message\SmsNotification: async_2
+
+.. code-block::php
+
+    <?php
```
Comment (OskarStark) [inline messenger.rst]:
```suggestion
```

**Comment by sylfabre** [comment]
@OskarStark thank you for your review.

Anything else I should do to get it merged?

File: messenger.rst [hunk 4]
```diff
@@ -431,6 +431,33 @@ transport and "sending" messages there to be handled immediately:
...
+Message Inheritance & Routing Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With the following code and configuration, messages will be dispatched like this:
+* ``Notification`` to ``async_1``
+* ``SmsNotification`` to ``async_1``
+* ``SmsNotification`` to ``async_2``
+
+.. code-block::php
+
```
Comment (Matts) [inline messenger.rst]:
```suggestion
```

File: messenger.rst [hunk 5]
```diff
@@ -431,6 +431,33 @@ transport and "sending" messages there to be handled immediately:
...
+    namespace App\Message;
+    class SmsNotification extends Notification {}
+
+.. code-block::yaml
+
```
Comment (Matts) [inline messenger.rst]:
```suggestion
```

File: messenger.rst [hunk 6]
```diff
@@ -431,6 +431,33 @@ transport and "sending" messages there to be handled immediately:
...
+With the following code and configuration, messages will be dispatched like this:
+* ``Notification`` to ``async_1``
+* ``SmsNotification`` to ``async_1``
+* ``SmsNotification`` to ``async_2``
+
+.. code-block::php
+
+    namespace App\Message;
+    class SmsNotification extends Notification {}
```
Comment (Matts) [inline messenger.rst]:
```suggestion

    class SmsNotification extends Notification {}
```

File: messenger.rst [hunk 7]
```diff
@@ -431,6 +431,33 @@ transport and "sending" messages there to be handled immediately:
...
+
+    # config/packages/messenger.yaml
+    framework:
+        messenger:
+            routing:
+                App\Message\Notification: async_1
+                App\Message\SmsNotification: async_2
+
+.. code-block::php
+
```
Comment (Matts) [inline messenger.rst]:
```suggestion
```

**Review by Matts — CHANGES_REQUESTED** [review]
@sylfabre After these changes LTGM

**Comment by sylfabre** [comment]
@Matts done, thank you for your review

File: messenger.rst [hunk 8]
```diff
@@ -431,6 +431,31 @@ transport and "sending" messages there to be handled immediately:
...
 
+Message Inheritance & Routing Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With the following code and configuration, messages will be dispatched like this:
+* ``Notification`` to ``async_1``
+* ``SmsNotification`` to ``async_1``
+* ``SmsNotification`` to ``async_2``
+
+.. code-block::php
```
Comment (Matts) [inline messenger.rst]:
```suggestion
.. code-block::php

```

File: messenger.rst [hunk 9]
```diff
@@ -431,6 +431,31 @@ transport and "sending" messages there to be handled immediately:
...
+    namespace App\Message;
+
+    class SmsNotification extends Notification {}
+
+.. code-block::yaml
```
Comment (Matts) [inline messenger.rst]:
```suggestion
.. code-block::yaml

```

File: messenger.rst [hunk 10]
```diff
@@ -431,6 +431,31 @@ transport and "sending" messages there to be handled immediately:
...
+    # config/packages/messenger.yaml
+    framework:
+        messenger:
+            routing:
+                App\Message\Notification: async_1
+                App\Message\SmsNotification: async_2
+
+.. code-block::php
```
Comment (Matts) [inline messenger.rst]:
```suggestion
.. code-block::php

```

**Review by Matts — APPROVED** [review]
I am so sorry! It seems the CI does not agree with my interpretation. Did not know that an empty line after a start of the code block is a requirement. See the required changes below. Will approve it now seeing as this is blocked by CI

**Comment by sylfabre** [comment]
@Matts no worries!

**Review by maxhelias — APPROVED** [review]

**Comment by sylfabre** [comment]
I don't why I closed it a month ago ...

**Comment by wouterj** [comment]
Thank you @sylfabre for the research & documentation. I've slightly reduced the content and moved it up a bit (as we already talk a bit about inheritance at the start of the section).

**Pull request merged**